product-service/handlers/catalog-batch-process.ts

```diff
@@ -0,0 +1,34 @@
+import { ProductDTO } from './../product.model';
+import { SQSEvent, SQSHandler } from 'aws-lambda';
+import { SNS } from 'aws-sdk';
+import { insertProduct } from '../db/insert-product';
+
+export const catalogBatchProcess: SQSHandler = (event: SQSEvent): void => {
+  console.log('Lambda invocation with event: ', event);
+
+  const sns = new SNS({ region: 'eu-west-1' });
+  const products: ProductDTO[] = event.Records.map(({ body }) => JSON.parse(body));
+
+  try {
+    products.forEach(async (p) => {
+      await insertProduct(p);
+    });
+    publishNotification(sns, products);
```

Comment:
`products.forEach()` does not wait for all the promises to resolve, it does not care if your callbacks are async. So `await` inside of the callback does not make any difference. As a consequence, `publishNotification` will not wait for all products to be inserted to the database, and it may happen so that the user will receive a notification saying that everything is OK, but in fact some of the insertions will fail.
If such a behavior is not desirable, you can solve this problem in one of the following ways:
 - replace `forEach()` with `for (... of ...)` and make your handler `async`
 - replace `forEach()` with `map()` and then `await` for `Promise.all()` (your handler still has to be `async`)

Reply:
Great catch! Will use `Promise.all()` instead